### DISCUSSION TITLE
Hashed admin token not replacing plain text token

### DISCUSSION THREAD

**Discussion by tommyalatalo** [dani-garcia/vaultwarden#3378] [Q&A]
### Subject of the issue
After updating to 1.28.0 and hashing the admin token with argon2 I keep getting a notice in the startup logs:

```
[NOTICE] You are using a plain text `ADMIN_TOKEN` which is insecure.
```

It seems that the plain text admin token is stored in vaultwarden internally and is not being overwritten when replacing `ADMIN_TOKEN` with a hashed value.

### Deployment environment

```
    ### Your environment (Generated via diagnostics page)
    * Vaultwarden version: v1.28.0
    * Web-vault version: v2023.3.0b
    * OS/Arch: linux/x86_64
    * Running within Docker: true (Base: Alpine)
    * Environment settings overridden: true
    * Uses a reverse proxy: true
    * IP Header check: true (X-Forwarded-For)
    * Internet access: true
    * Internet access via a proxy: false
    * DNS Check: true
    * Browser/Server Time Check: true
    * Server/NTP Time Check: true
    * Domain Configuration Check: true
    * HTTPS Check: true
    * Database type: PostgreSQL
    * Database version: PostgreSQL 9.6.24 on x86_64-pc-linux-musl, compiled by gcc (Alpine 10.3.1_git20211027) 10.3.1 20211027, 64-bit
    * Clients used: 
    * Reverse proxy and version: 
    * Other relevant information: 

**Environment settings which are overridden:** DOMAIN, SIGNUPS_ALLOWED, INVITATIONS_ALLOWED, SHOW_PASSWORD_HINT, ADMIN_TOKEN, IP_HEADER, DUO_IKEY, DUO_SKEY, DUO_HOST, SMTP_HOST, SMTP_PORT, SMTP_FROM, SMTP_FROM_NAME, SMTP_USERNAME, SMTP_PASSWORD
````

* Install method: Docker

### Steps to reproduce

1. Set up Vaultwarden with plain text admin token
2. After Vaultwarden is up and running with plain text token, generate a hashed token with argon2
3. Change `ADMIN_TOKEN` to hashed token and restart Vaultwarden
4. Vaultwarden logs warning about using a plain text token with `ADMIN_TOKEN` set to a hashed token

### Expected behaviour
Vaultwarden would take the hashed token and be happy, at least if the unhashed value is the actual password value.

### Actual behaviour
Vaultwarden logs incorrect information about configured ADMIN_TOKEN values

**Comment by BlackDex** [comment]
You used the admin interface to save settings. This creates a `config.json` file which overrules all ENV variables.
As is stated both during startup and in the admin interface at the top.

Either login into the admin interface, and replace the token there and press save. Or make sure all changed settings are converted to ENV variables and remove the `config.json`.

**Reply by jpaechnatz** [reply]
I did the same.
Afterwards, I see the "unescaped" (no double $$)  $argon2id$ entry in the config.json. The wiki page does not state how to login after changing it to argon2,  which part of the string? Salt, hash, complete? Or use the Password which was hashed by argon2?

**Reply by BlackDex** [reply]
Well, it would be strange if you would use that argon2 string it self not?
You created that argon2 string providing a password, so that is what should be used.

**Reply by Wever1985** [reply]
> Well, it would be strange if you would use that argon2 string it self not? You created that argon2 string providing a password, so that is what should be used.

That does make sense but I have to admit I was confused as well until reading this reply. The input field on the admin page literally says "Enter admin token" not "password". I think it wouldn't hurt to add a clarifying sentence to the wiki page.

**Reply by andymarden** [reply]
Oh good grief? Thought I was going mad. When you set ADMIN_TOKEN in the env file and then on the Admin page you get asked to enter a token it's kinda expected they mean the same thing. That web page should really be changed as suggested.

**Reply by BlackDex** [reply]
It would be rather strange to convert a password input into a hash and then literally input that hash as a password of course. 

Then hashing it would be rather stupid to do, and still providing and storing that in plain text.

**Reply by andymarden** [reply]
Perhaps but it doesn't alter the fact that the instruction is just plain misleading. In fact, this says it is just a token and tokens are normally long strings of nonsense.

**Comment by marcogiorgio** [comment]
Hi I have the same problem but I didn't override anything through the `config.json` . If it's better to open another issue please let me know.
What I did:
- Enter into Vaultwarden docker bash shell and execute: `/vaultwarden hash`
- Escape the `$` characters as in order to obtain a string like this `$$argon2id$$v=19$$m=65540,t=3,p=4$$VPWgVtNL` (the complete string is truncated)
- set the `ADMIN_TOKEN` variable like this: `ADMIN_TOKEN=$$argon2id$$v=19$$m=65540,t=3,p=4$$VPWgVtNL`
-  Recreate the docker container

I still get this error in the console:
<details>
  <summary>Error</summary>
 
  
  ```bash
 [NOTICE] You are using a plain text `ADMIN_TOKEN` which is insecure.

Please generate a secure Argon2 PHC string by using `vaultwarden hash` or `argon2`.

See: https://github.com/dani-garcia/vaultwarden/wiki/Enabling-admin-page#secure-the-admin_token

[2023-03-27 09:42:20.886][vaultwarden::api::notifications][INFO] Starting WebSockets server on 0.0.0.0:3012
[2023-03-27 09:42:20.888][start][INFO] Rocket has launched from http://0.0.0.0:9070
  ```
  
</details>

**Reply by BlackDex** [reply]
Are you sure you do not have a `config.json`?
Also, depending on where you configure the token, you maybe do not need to use `$$` and a single `$` could be enough.

For example, i use a `.env` file for my `docker-compose.yaml`, and i use the variables from the `.env` file within my `docker-compose.yaml` to configure the variable. In this situation i do not need to use `$$` in the `.env` file.

Same btw when using a `.env` file directly for Vaultwarden without the usage of docker.
Or, when using it via the CLI, like `docker run -e ADMIN_TOKEN='$argon2id$v....' vaultwarden/server` on those locations the double `$$` will not work.

**Reply by marcogiorgio** [reply]
Aw this is my case also:

>For example, i use a .env file for my docker-compose.yaml, and i use the variables from the .env file within my docker-compose.yaml to configure the variable. In this situation i do not need to use $$ in the .env file.

I removed the escaped `$` and now all is working properly, thanks! Maybe it would be useful to insert this note in the docs :smiley:

**Reply by BlackDex** [reply]
I have updated the docs:
https://github.com/dani-garcia/vaultwarden/wiki/Enabling-admin-page#using-vaultwarden-hash

**Comment by jpaechnatz** [comment]
And please add the Advice, the answer to my question ;-), after changing to the argon2 hash, please login with the password what was used for hashing, althought the admin webpage still says "token".

**Reply by BlackDex** [reply]
token/password in this case is a synonym. Previously i wouldn't even call it a token either, since we used it as a password also.
But changing that would be breaking current installs, and that's not what we want.

**Comment by tommyalatalo** [comment]
Thanks, in my case I had the `config.json` file but when checking all of the values that we're non-default in the UI were also set by my env variables, which was also my expectation, so removing the config.json file was no problem.

**Reply by tommyalatalo** [reply]
i have a follow up issue with this. After removing the config.json file and restarting vaultwarden (multiple times), the Duo global `enable` setting is never set to enabled even though I have all three of `DUO_HOST`, `DUO_IKEY` and `DUO_SKEY` set. I've checked several times that I don't have a config.json file in my container, there definitely isn't. 

In order to enable Duo now I seem to have no choice but to tick the `enable` box in the UI and save, which creates the config.json file that I want to avoid. Am I missing something to enable the Duo integration automatically?

**Reply by BlackDex** [reply]
You can also set `_ENABLE_DUO=true` as env.

**Reply by tommyalatalo** [reply]
Yeah, I was looking for something like that, which wasn't documented.
But it does seem like there might be something wrong with the Duo enabling though, I'm quite sure that my configuration is correct.

**Reply by BlackDex** [reply]
Nothing has changed there for years i think

**Reply by tommyalatalo** [reply]
```
DUO_IKEY=DI2....
DUO_SKEY=1tZ....
DUO_HOST=api-9...duosecurity.com
```

The above are all set in my container env, but the Duo integration doesn't get enabled without setting `_ENABLE_DUO`  like you mentioned. Unless there is some other state that's stored on disk for this I would say that it's likely broken and needs a look.

I can open an issue for this.

**Reply by BlackDex** [reply]
No that is exactly correct. You need to add that flag, else it will not work.

**Reply by stefan0xC** [reply]
@BlackDex Seems weird that users have to set an internal variable.  Shouldn't we set `_enable_duo` dependent on whether or not the other settings are set?

**Reply by BlackDex** [reply]
I'm not historically sure why this is like this. And we may want to improve it. But it isn't something changed in the past releases.
I actually never ever have configured duo for Vaultwarden my self yet.

**Reply by stefan0xC** [reply]
Me neither. :wink:
@altosys you might want to create a separate issue for this so this gets more attention. Otherwise I'd suggest that you can maybe create a wiki page to help others with setting up duo?

**Reply by BlackDex** [reply]
I see that 4 years ago @dani-garcia replaced true with false

**Reply by tommyalatalo** [reply]
Issue for the Duo integration: https://github.com/dani-garcia/vaultwarden/issues/3382
I guess we'll continue discussion over there, but it does seem to me that the conditional controlling the Duo integration is wrong.
